src/data-access/services/round-robin-ftp-client-pool.ts

```diff
@@ -0,0 +1,113 @@
+import { ConnectionStatus } from '../enums/connection-status'
+import { FtpClient } from '../interfaces/ftp-client'
+import { FileMetadata } from '../value-objects/file-metadata'
+
+const MAX_FAILED_OPERATIONS_THRESHOLD: number = 20
+const SUCCESSFUL_OPERATION_WEIGHT: number = 1
+const FAILED_OPERATION_WEIGHT: number = 2
+
+export class RoundRobinFtpClientPool implements FtpClient {
+  private connectedFtpClient?: FtpClient
+  private onConnectionStatusChangedCallback?: (connectionStatus: ConnectionStatus) => void
+  private failedOperationsTracker: number = 0
+
+  public constructor(private ftpClients: readonly FtpClient[]) {}
+
+  public async connect(): Promise<void> {
+    await this.getConnectedFtpClient()
+  }
+
+  private async getConnectedFtpClient(): Promise<FtpClient> {
+    if (!this.connectedFtpClient || !this.connectedFtpClient.isConnected()) {
+      this.connectedFtpClient = await this.getFtpClientByRoundRobin()
+    }
+    return this.connectedFtpClient
+  }
+
+  private async getFtpClientByRoundRobin(): Promise<FtpClient> {
+    await this.disconnect()
+    this.emitConnectionStatus(ConnectionStatus.CONNECTING)
+    for (const ftpClient of this.ftpClients) {
+      try {
+        await ftpClient.connect()
+        this.emitConnectionStatus(ConnectionStatus.CONNECTED)
+        ftpClient.setOnConnectionStatusChangedCallback(connectionStatus => this.emitConnectionStatus(connectionStatus))
+        return ftpClient
+      } catch {
+        ftpClient.setOnConnectionStatusChangedCallback(() => {})
+      }
+    }
```

Comment:
Might i suggest some spacings to make it easier to see the compartmented code? 😄 
```suggestion

    for (const ftpClient of this.ftpClients) {
      try {
        await ftpClient.connect()
        this.emitConnectionStatus(ConnectionStatus.CONNECTED)
        ftpClient.setOnConnectionStatusChangedCallback(connectionStatus => this.emitConnectionStatus(connectionStatus))
        return ftpClient
      } catch {
        ftpClient.setOnConnectionStatusChangedCallback(() => {})
      }
    }
    
```

Reply:
Updated.